edward/inferences/conjugacy/conjugacy.py

```diff
@@ -0,0 +1,246 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import re
+from collections import defaultdict
+from pprint import pprint
+from copy import copy
+
+import numpy as np
+import tensorflow as tf
+
+from edward.models.random_variable import RandomVariable
+from edward.models import random_variables as rvs
+
+import edward.inferences.conjugacy.conjugate_log_probs
+from edward.inferences.conjugacy.simplify \
+    import symbolic_suff_stat, full_simplify, expr_contains, reconstruct_expr
+
+copy_op_to_graph = tf.contrib.copy_graph.copy_op_to_graph
+
+# TODO(mhoffman): Support for slicing, tf.gather, etc.
+
+
+def normal_from_natural_params(p1, p2):
+  sigmasq = 0.5 * tf.reciprocal(-p1)
+  mu = sigmasq * p2
+  return {'mu': mu, 'sigma': tf.sqrt(sigmasq)}
+
+
+rvs.Bernoulli.support = 'binary'
+rvs.Categorical.support = 'onehot'
+rvs.Beta.support = '01'
+rvs.Dirichlet.support = 'simplex'
+rvs.Gamma.support = 'nonnegative'
+rvs.InverseGamma.support = 'nonnegative'
+rvs.Normal.support = 'real'
+
+
+_suff_stat_to_dist = defaultdict(dict)
+_suff_stat_to_dist['binary'][(('#x',),)] = (
+    rvs.Bernoulli, lambda p1: {'p': tf.sigmoid(p1)})
+_suff_stat_to_dist['onehot'][(('#OneHot', ('#x',),),)] = (
+    rvs.Categorical, lambda p1: {'p': tf.nn.softmax(p1)})
+_suff_stat_to_dist['01'][(('#Log', ('#One_minus', ('#x',))),
+                          ('#Log', ('#x',)))] = (
+    rvs.Beta, lambda p1, p2: {'a': p2 + 1, 'b': p1 + 1})
+_suff_stat_to_dist['simplex'][(('#Log', ('#x',)),)] = (
+    rvs.Dirichlet, lambda p1: {'alpha': p1 + 1})
+_suff_stat_to_dist['nonnegative'][(('#Log', ('#x',)),
+                                   ('#x',))] = (
+    rvs.Gamma, lambda p1, p2: {'alpha': p1 + 1, 'beta': -p2})
+_suff_stat_to_dist['nonnegative'][(('#CPow-1.0000e+00', ('#x',)),
+                                   ('#Log', ('#x',)))] = (
+    rvs.InverseGamma, lambda p1, p2: {'alpha': -p2 - 1, 'beta': -p1})
+_suff_stat_to_dist['real'][(('#CPow2.0000e+00', ('#x',)),
+                            ('#x',))] = (
+    rvs.Normal, normal_from_natural_params)
+
+
+def _log_joint_name(blanket):
+  return '_log_joint_of_' + ('&'.join([i.name[:-1] for i in blanket])) + '_'
+
+
+def get_log_joint(blanket):
```

Comment:
That's right. I introduced this caching for other reasons (which aren't all that relevant now), but it still speeds up graph construction.

But maybe it makes sense to move that caching a level up, so that each `RandomVariable` caches its `conjugate_log_prob()`?

(The term "cache" is a little funny here, since the graph remembers everything regardless. We're just caching a pointer.)